master/pkg/schemas/merge.go

```diff
@@ -0,0 +1,187 @@
+// is settable.
+//
+// The default behavior for merging maps is to merge keys from src to obj, and the default
+// behavior for slices is to copy them.  This is analgous to how json.Unmarshal treats maps and
+// slices.  However, the default merging behavior for an object can be overwritten by implementing
+// the Mergable interface.  An example of this is BindMountsConfig.
+//
+// Merge is intelligent enough to handle union types automatically.  In those cases, Merge is
+// recursive as log as the obj getting filled either does not have any of the union types defined
+// or if it has the same union type defined as the src.  That is, a S3 checkpoint storage object
+// will never be used as a src to try to fill a SharedFS checkpoint storage object.
+//
+// Example usage:
+//
+//    config, err := expconf.ParseAnyExperimentConfigYAML(bytes)
+//
+//    var cluster_default_checkpoint_storage expconf.CheckpointStorage = ...
+//
+//    // Use the cluster checkpoint storage if the user did not specify one.
+//    schemas.Merge(&config.CheckpointStorage, cluster_default_checkpoint_storage)
+//
+func Merge(obj interface{}, src interface{}) {
+	// fmt.Printf("--------------------\n")
+	name := fmt.Sprintf("%T", obj)
+	merge(reflect.ValueOf(obj), reflect.ValueOf(src), false, name)
+}
+
+func assertTypeMatch(obj reflect.Value, src reflect.Value) {
+	if obj.Type() == src.Type() {
+		return
+	}
+	panic(
+		fmt.Sprintf(
+			"type mismatch in merge; can't fill %T with %T",
+			obj.Interface(),
+			src.Interface(),
+		),
+	)
+}
+
+// merge is the recursive layer under Merge.
+func merge(obj reflect.Value, src reflect.Value, allocated bool, name string) {
+	// fmt.Printf("merge(%T, %T)\n", obj.Interface(), src.Interface())
+	// First deref src.  If it is ultimately nil, stop.
+	for src.Kind() == reflect.Ptr || src.Kind() == reflect.Interface {
+		if src.IsZero() {
+			return
+		}
+		// Recurse with dereferenced src.
+		merge(obj, src.Elem(), allocated, name)
+		return
+	}
+
+	// If the object is Mergable, we only call Merge and nothing else.
+	if mergeable, ok := obj.Interface().(Mergable); ok {
+		mergeable.Merge(src.Interface())
+		return
+	}
+
+	// obj should always be a pointer, because Merge(&x, y) will act on x in-place.
+	if obj.Kind() != reflect.Ptr {
+		panic("non-pointer in merge")
+	}
+	// obj can't be a nil pointer, because Merge(nil, y) doesn't make any sense.
+	if obj.IsZero() {
+		panic("nil pointer in merge")
+	}
+	// Now operate on what obj points to.
+	obj = obj.Elem()
+
+	switch obj.Kind() {
+	case reflect.Interface:
+		if obj.IsZero() {
+			// This doesn't make any sense; we need a type.
+			panic("got a nil interface as the obj to merge into")
+		}
+		// Dereference the type but not the original pointer.
+		merge(obj.Elem().Addr(), src, allocated, name)
+
+	case reflect.Ptr:
+		// Note that this is a double-pointer since we already dereference the object once.
+		if obj.IsZero() {
+			// Allocate, then recurse with allocated = true.
```

Comment:
I'm a bit confused. Wouldn't a pointer to a nil pointer be just as unsettable as being given a pointer? Also do we even need to handle the double pointer case? What does `obj.Set(tmp)` even do here?

Reply:
It seems to me after some thought should just `derefInput` it and run the struct/map/slice cases.